### PR TITLE
Refactor gRPC running in platform mode to ir

### PR DESCRIPTION
**Pull request by dianadoherty**
https://github.com/meroxa/turbine-core/issues/7

TODO:
- [x] Save IR information when running gRPC server in platform mode (all endpoints should work in this mode)
- [x] Remove dependency on platform API
- [x] Ensure that function information is being recorded
- [x] Validate the IR spec 
- [ ] Test, test, test 

### Testing Steps
**Setup**
- copy an example ruby app from 
- [pull cli branch for testing ir work](https://github.com/meroxa/cli/compare/temp-ruby-branch?expand=1#diff-a3040979337a1427de0c15c19d24f1667a35559b04d1765e2b4007584ad20e89R17). Note that you can switch from different gRPC modes for testing https://github.com/meroxa/cli/compare/temp-ruby-branch?expand=1#diff-a3040979337a1427de0c15c19d24f1667a35559b04d1765e2b4007584ad20e89R17 

**Testing changes to turbine ruby**
- from `turbine-core/lib/ruby/turbine` run `gem build turbine.gemspec`
- from `simple` test application, pull latest local gem: `gem install --local '/Users/dianadoherty/go/src/meroxa/turbine-core/lib/ruby/turbine/turbine-0.1.0.gem'`
- from `simple` test application, run `meroxa apps run`

**Testing changes to turbine core**
- from `cli` run `make gomod && make install`
- from `simple` test application, run `meroxa apps run`

Current IR spec:
Note that function information is not currently being saved
```
ir.DeploymentSpec{
  Secrets:map[string]string(nil), 
  Connectors:[]ir.ConnectorSpec{
    ir.ConnectorSpec{
      Type:"source", 
      Resource:"demopg", 
      Collection:"events", 
      Config:map[string]interface {}{}
    }, 
    ir.ConnectorSpec{
      Type:"destination", 
      Resource:"demopg", 
      Collection:"events_copy",
      Config:map[string]interface {}{}
    }
  }, 
  Functions:[]ir.FunctionSpec(nil), 
  Definition:ir.DefinitionSpec{
    GitSha:"", 
    Metadata:ir.MetadataSpec{
      Turbine:ir.TurbineSpec{
        Language:"RUBY",
        Version:"0.1.0"
      }, 
      SpecVersion:"0.1.1"
      }
   }
}
```

### CODE REVIEW

**Comment by raulb** [comment]
For the record, @lyuboxa is tackling https://github.com/meroxa/turbine-core/issues/8 on this pull-request.

File: cmd/server/main.go [hunk 1]
```diff
@@ -20,22 +16,20 @@ var (
 )
 
 func main() {
-	flag.StringVar(&Mode, "mode", "info", "gRPC server mode. Options are info, platform and local. Default is info.")
+	flag.StringVar(&Mode, "mode", "record", "gRPC server mode. Options are record and run. Default is record.")
```
Comment (lyuboxa) [inline cmd/server/main.go:19]: I don't think we need main, we should be calling the grpc server in the command line.

File: lib/ruby/turbine/lib/turbine.rb [hunk 2]
```diff
@@ -10,25 +10,38 @@ class Error < StandardError; end
   class Runner
```
Comment (lyuboxa) [inline lib/ruby/turbine/lib/turbine.rb:10]: I don't think anything in `lib/ruby` is needed?

File: pkg/server/record.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	pb "github.com/meroxa/turbine-core/lib/go/github.com/meroxa/turbine/core"
+	"github.com/meroxa/turbine-core/pkg/ir"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type recordService struct {
```
Comment (lyuboxa) [inline pkg/server/record.go:14]: I realized that `record` as noun can be ambiguous, I think `recordingService` is likely more descriptive.. but maybe not.

File: pkg/server/record.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+func NewRecordService() *recordService {
+	return &recordService{}
+}
+
+func (s *recordService) Init(ctx context.Context, request *pb.InitRequest) (*emptypb.Empty, error) {
+	s.deploymentSpec.Definition = ir.DefinitionSpec{
+		GitSha: request.GetGitSHA(),
+		Metadata: ir.MetadataSpec{
+			Turbine: ir.TurbineSpec{
+				Language: ir.Lang(strings.ToLower(request.GetLanguage().String())),
```
Comment (lyuboxa) [inline pkg/server/record.go:29]: Do we want to validate `Lang`, right now it will just cast any string to that type?

File: pkg/server/record.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		ir.ConnectorSpec{
+			Collection: request.GetTargetCollection(),
+			Resource:   request.GetResource().GetName(),
+			Type:       ir.ConnectorDestination,
+			Config:     resourceConfigsToMap(request.GetConfigs().GetResourceConfig()),
+		},
+	)
+
+	b, err := json.Marshal(s.deploymentSpec)
+	fmt.Printf("MARSHAL ERR: %#+v\n", err)
```
Comment (lyuboxa) [inline pkg/server/record.go:98]: debug?

File: pkg/server/server.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
+	}
+
+	if err := s.Serve(listener); err != nil {
+		log.Fatalf("failed to serve: %v", err)
+	}
+}
+
+func Empty() *emptypb.Empty {
```
Comment (lyuboxa) [inline pkg/server/server.go:45]: nit; this doesn't have to be exportable

**Review by lyuboxa — APPROVED** [review]
Ship it.

File: lib/ruby/turbine/lib/turbine.rb (hunk 2 above)
Comment (raulb) [reply lib/ruby/turbine/lib/turbine.rb:10]: @lyuboxa not sure I understand. Isn't there where the implementation of ruby is taking place? Maybe you mean that could be nested differently?

File: pkg/server/record.go (hunk 3 above)
Comment (raulb) [reply pkg/server/record.go:14]: @lyuboxa don't have a particular opinion. Maybe we could leave this as is for now and iterate if we need to.

File: pkg/server/record.go (hunk 5 above)
Comment (raulb) [reply pkg/server/record.go]: https://github.com/meroxa/turbine-core/pull/13/commits/8b041eb318d1b978b968c52d60c105828200e9b9

File: pkg/server/server.go (hunk 6 above)
Comment (raulb) [reply pkg/server/server.go]: https://github.com/meroxa/turbine-core/pull/13/commits/1e3af3506949ab067ccfbc8e1d6b00171237cec7

**Review by lyuboxa — APPROVED** [review]

**Pull request merged**